File: 876-middle-node/Solution.java
class ListNode{
    int val;
    ListNode next;
    ListNode prev;
    ListNode(int val,ListNode prev, ListNode next){
        this.val = val;
        this.prev = prev;
        this.next = next;
    }
}
class LinkedList{
    ListNode head;
    ListNode tail;
    void AddAtTail(ListNode newNode){
        if (head == null){
            head = newNode;
            tail = newNode;
            return;
        }
        newNode.prev = tail;
        tail.next = newNode;
        tail = newNode;
    }
}
class Solution876 {
    public ListNode middleNode(ListNode head) {
        if (head == null){
            return null;
        }
        ListNode dummy = new ListNode(0, null, head);
        ListNode cur = dummy;
        int size = 0;
        while (cur.next != null){
            cur = cur.next;
            size++;
            System.out.println(size);
        }
        cur = dummy;
        int index = -1;
        while (cur.next != null){
            cur = cur.next;
            index++;
            System.out.printf("cur.val = %s ", cur.val);
            System.out.printf("index = %s%n", index);
            if (index >= size/2){
                System.out.printf("return cur %s", cur.val);
                break;
            }
        }
        return cur;
    }

    static void main(String[] args) {
        int[] nums = {1, 2, 3, 4, 5};
        int[] nums2 = {1};
        LinkedList mainList = new LinkedList();
        for (int i : nums2){
            ListNode newNode = new ListNode(i, null, null);
            mainList.AddAtTail(newNode);
        }
        Solution876 sol = new Solution876();
        sol.middleNode(mainList.head);
    }
}

File: 876-middle-node/solution.go
package main

type ListNode struct {
	val  int
	Next *ListNode
}
type LinkedList struct {
	head *ListNode
	tail *ListNode
	size int
}

func (ll *LinkedList) Append(val int) {
	node := &ListNode{val, nil}
	if ll.head == nil {
		ll.head = node
		ll.tail = node
	} else {
		ll.tail.Next = node
		ll.tail = node
	}
	ll.size++
}

func middleNode(head *ListNode) *ListNode {
	i, k := head, head
	size := 1
	for i.Next != nil {
		i = i.Next
		size++
	}
	for j := 0; j < size/2; j++ {
		k = k.Next
	}
	return k
}
func main() {
	head := []int{1}
	ll := LinkedList{}
	for _, v := range head {
		ll.Append(v)
	}
	middleNode(ll.head)
}

//思路: 第一次遍歷先得到總長度, 第二次遍歷時走到一半就直接 return
